File: CsLibrary/CsLibrary/Short.cs
using System;

namespace JavaClasses
{

    /// <summary>
    /// The Short class wraps a value of primitive type short in an object.
    /// </summary>
    /// <author>Petr Kozler (A13B0359P)</author>
    public class Short : Comparable<Short>
    {

        /// <summary>
        /// A constant holding the maximum value a short can have, 2^15-1.
        /// </summary>
        public const short MAX_VALUE = 32767;

        /// <summary>
        /// A constant holding the minimum value a short can have, -2^15.
        /// </summary>
        public const short MIN_VALUE = -32768;

        private short v;

        /// <summary>
        /// Constructs a newly allocated Short object that represents the specified
        /// short value.
        /// </summary>
        /// <param name="value">the value to be represented by the Short.
        /// </param>
        public Short(short value)
        {
            v = value;
        }

        /// <summary>
        /// Returns the value of this Short as a short.
        /// </summary>
        /// <returns>the numeric value represented by this object after conversion to
        /// type short.
        /// </returns>
        public short shortValue()
        {
            return v;
        }

        /// <summary>
        /// Compares two Short objects numerically.
        /// </summary>
        /// <param name="anotherShort">the Short to be compared.
        /// </param><returns>the value 0 if this Short is equal to the argument Short; a value
        /// less than 0 if this Short is numerically less than the argument Short;
        /// and a value greater than 0 if this Short is numerically greater than the
        /// argument Short (signed comparison).
        /// </returns>
        public override int compareTo(Short anotherShort)
        {
            return compare(v, anotherShort.v);
        }

        /// <summary>
        /// Compares two short values numerically.
        /// </summary>
        /// <param name="x">the first short to compare
        /// </param><param name="y">the second short to compare
        /// </param><returns>the value 0 if x == y; a value less than 0 if x < y; and a value greater than 0 if x
        /// > y
        /// </returns>
        public static int compare(short x, short y)
        {
            return x - y;
        }

        /// <summary>
        /// Compares this object to the specified object.
        /// </summary>
        /// <param name="obj">the object to compare with
        /// </param><returns>true if the objects are the same; false otherwise.
        /// </returns>
        public override bool equals(Object obj)
        {
            if (obj == null)
            {
                return false;
            }

            if (GetType() != obj.GetType())
            {
                return false;
            }

            return v.Equals((obj as Short).v);
        }

        /// <summary>
        /// Returns a String object representing this Short's value.
        /// </summary>
        /// <returns>a string representation of the value of this object in base 10.
        /// </returns>
        public override String toString()
        {
            return Short.toString(v);
        }

        /// <summary>
        /// Returns a new String object representing the specified short.
        /// </summary>
        /// <param name="s">the short to be converted
        /// </param><returns>the string representation of the specified short
        /// </returns>
        public static String toString(short s)
        {
            return new String(s.ToString());
        }

        /// <summary>
        /// Parses the string argument as a signed decimal short.
        /// </summary>
        /// <param name="s">a String containing the short representation to be parsed
        /// </param><returns>the short value represented by the argument in decimal.
        /// </returns>
        public static short parseShort(String s)
        {
            return short.Parse(s.ToString());
        }
        
        public static implicit operator Short(short original)
        {
            return new Short(original);
        }
        
    }

}

File: CppLibrary/Short.cpp
#include "Short.h"

#include <string>

/**
 * A constant holding the maximum value a short can have, 2^15-1.
 */
const int16_t Short::MAX_VALUE = 32767;
/**
 * A constant holding the minimum value a short can have, -2^15.
 */
const int16_t Short::MIN_VALUE = -32768;

/**
 * Constructs a newly allocated Short object that represents the specified
 * short value.
 *
 * @param value the value to be represented by the Short.
 */
Short::Short(int16_t value) {
    v = value;
}

/**
 * Returns the value of this Short as a short.
 *
 * @return the numeric value represented by this object after conversion to
 * type short.
 */
int16_t Short::shortValue() {
    return v;
}

/**
 * Compares two Short objects numerically.
 *
 * @param anotherShort the Short to be compared.
 * @return the value 0 if this Short is equal to the argument Short; a value
 * less than 0 if this Short is numerically less than the argument Short;
 * and a value greater than 0 if this Short is numerically greater than the
 * argument Short (signed comparison).
 */
int32_t Short::compareTo(Short *anotherShort) {
    return compare(v, anotherShort->v);
}

/**
 * Compares two short values numerically.
 *
 * @param x the first short to compare
 * @param y the second short to compare
 * @return the value 0 if x == y; a value less than 0 if x < y; and a value greater than 0 if x
 * > y
 */
int32_t Short::compare(int16_t x, int16_t y) {
    return x - y;
}

/**
 * Compares this object to the specified object.
 *
 * @param obj the object to compare with
 * @return true if the objects are the same; false otherwise.
 */
int16_t Short::equals(Short *obj) {
    if (obj == nullptr) {
        return false;
    }

    if (sizeof(*this) != sizeof(*obj)) {
        return false;
    }
    
    return (v == obj->v);
}

/**
 * Returns a String object representing this Short's value.
 *
 * @return a string representation of the value of this object in base 10.
 */
String *Short::toString() {
    return toString(v);
}

/**
 * Returns a new String object representing the specified short.
 *
 * @param s the short to be converted
 * @return the string representation of the specified short
 */
String *Short::toString(int16_t s) {
    return new String(std::to_string(s));
}

/**
 * Parses the string argument as a signed decimal short.
 *
 * @param s a String containing the short representation to be parsed
 * @return the short value represented by the argument in decimal.
 */
int16_t Short::parseShort(String *s) {
    return (int16_t) std::stoi(s->toString(), nullptr, 10);
}

std::ostream &operator<<(std::ostream &s, Short &obj) {
    return s << obj.toString()->toString().c_str();
}
